src/node/mini_miner.cpp

```diff
@@ -0,0 +1,265 @@
+        const auto ancestor_package_size = (*best_iter)->second.GetSizeWithAncestors();
+        const auto ancestor_package_fee = (*best_iter)->second.GetModFeesWithAncestors();
+        // Stop here. Everything that didn't "make it into the block" has bumpfee.
+        if (best_iter == entries.end() || ancestor_package_fee < target_feerate.GetFee(ancestor_package_size)) {
+            break;
+        }
+
+        // Calculate ancestors on the fly. This lookup should be fairly cheap, and ancestor sets
+        // change at every iteration, so this is more efficient than maintaining a cache.
+        std::set<MockEntryMap::iterator, IteratorComparator> ancestors;
+        std::set<MockEntryMap::iterator, IteratorComparator> to_process;
+        ancestors.insert(*best_iter);
+        to_process.insert(*best_iter);
+        while (!to_process.empty()) {
+            auto iter = to_process.begin();
+            assert(iter != to_process.end());
+            const CTransaction& tx = (*iter)->second.GetTx();
+            for (const auto& input : tx.vin) {
+                if (auto parent_it{entries_by_txid.find(input.prevout.hash)}; parent_it != entries_by_txid.end()) {
+                    to_process.insert(parent_it);
+                    ancestors.insert(parent_it);
+                }
+            }
+            to_process.erase(iter);
+        }
+        assert(ancestor_package_size == std::accumulate(ancestors.cbegin(), ancestors.cend(), 0,
+            [](int64_t sum, const auto it) {return sum + it->second.GetTxSize();}));
+        assert(ancestor_package_fee == std::accumulate(ancestors.cbegin(), ancestors.cend(), 0,
+            [](CAmount sum, const auto it) {return sum + it->second.GetModifiedFee();}));
+
+        // "Mine" all transactions in this ancestor set.
+        for (const auto& anc : ancestors) {
+            in_block.insert(anc->second.GetTx().GetHash());
+            total_fees += anc->second.GetModifiedFee();
+            total_vsize += anc->second.GetTxSize();
+            auto it = descendant_set_by_txid.find(anc->second.GetTx().GetHash());
+            assert(it != descendant_set_by_txid.end());
+            for (const auto& descendant : it->second) {
+                descendant->second.vsize_with_ancestors -= anc->second.GetTxSize();
+                descendant->second.fee_with_ancestors -= anc->second.GetModifiedFee();
+            }
+        }
+        DeleteAncestorPackage(ancestors);
+        assert(entries.size() == entries_by_txid.size());
+    }
+}
+
+std::map<COutPoint, CAmount> MiniMiner::CalculateBumpFees(const CFeeRate& target_feerate)
+{
+    // Build a block template until the target feerate is hit.
+    BuildMockTemplate(target_feerate);
+    assert(in_block.empty() || CFeeRate(total_fees, total_vsize) >= target_feerate);
+
+    // Each transaction that "made it into the block" has a bumpfee of 0, i.e. they are part of an
+    // ancestor package that exceeds the target feerate and don't need to be bumped.
```

Comment:
```suggestion
    // ancestor package with at least the target feerate and don't need to be bumped.
```
(if the feerates are equal, no bump is required)